vumi/persist/fields.py

```diff
@@ -957,3 +963,27 @@ class ManyToMany(ForeignKey):
 
     def __init__(self, other_model, index=None, backlink=None):
         super(ManyToMany, self).__init__(other_model, index, backlink)
+
+
+class ComputedIndexDescriptor(FieldDescriptor):
+    """A field descriptor for computed index fields."""
+
+    def pre_save(self, modelobj):
+        self.set_value(modelobj, self.field.value_func(modelobj))
+
+    def _add_index(self, modelobj, value):
+        # We override `FieldDescriptor._add_index` because we don't want `None`
+        # values.
+        if value is not None:
+            modelobj._riak_object.add_index(self.index_name, str(value))
+
+
+class ComputedIndex(Field):
+    """Field that stores a computed value for indexing."""
+
+    descriptor_class = ComputedIndexDescriptor
+
+    def __init__(self, value_func, index=True, **kw):
+        kw["null"] = True
+        super(ComputedIndex, self).__init__(index=index, **kw)
```

Comment:
We always need `null=True` so we can create models without having values in the fields yet.

I'm not sure if `index=False` really makes sense. It doesn't in the current implementation, but I've been wondering if it's worth doing the computation whenever any field value changes. I've been leaning toward doing that, in which case this becomes a `ComputedValue` field that may or may not be indexed rather than a `ComputedIndex` field.